### PR TITLE
Implement full IML plugin stack

### PR DESCRIPTION
**Pull request by jgrund**
This patch implements the full iml plugin stack including the agent and
IML manager sides.

Both the agent and manager sides are implemented here for ease of
maintenance.

This still avoids a monolithic situation, as separate areas of
responsibility are organized into Crates under a single
[workspace](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html).

This is now more akin to a modulith, where module specific concerns and
boundaries are maintained under a single repo.

All networked components are implemented with [Tokio](https://tokio.rs/)
to address any speed / scaling / code maintainability concerns.

  - Add Makefile + spec to build the whole cargo workspace.
  - Add section to `chroma-manager.conf` to consume agent messages in the
manager
  - Add RabbitMQ settings to shared IML manager environment.
  - Add Crate for manager-side services.
  - Add Crate to send / receive agent messages to IML manager.
  - Add Crate to run daemon / action plugins on IML agents.
  - Add Crate for proper CLI capabilities on IML agent.
  - Add a few simple reference plugins to iml-agent and iml-agent-comms crates.
  - Add a Crate to cleanly shutdown tokio runtimes on error.
  - Some small Python fixups.

Create an iml-comms consumer that somewhat mirrors chroma_agent_comms, but is in Rust.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

### CODE REVIEW

**Comment by jgrund** [comment]
image

File: python-iml-manager.spec.in [hunk 1]
```diff
@@ -34,7 +34,9 @@ Source11:       iml-power-control.service
 Source12:       iml-settings-populator.service
```
Comment (johnsonw) [inline python-iml-manager.spec.in]: Maybe in another patch we can update the repo to use the new copr deploy method, which should no longer require the spec.in file (and `includes` directory of course).

File: .copr/Makefile [hunk 2]
```diff
@@ -0,0 +1,15 @@
+srpm:
+	rm -rf _topdir *.rpm
+	rm -rf target/release/rust-iml
+	mkdir -p _topdir/SOURCES/
+	cargo build --release
+	mkdir -p target/release/rust-iml
+	cp target/release/iml-{action-runner,agent,agent-daemon,agent-comms,agent-daemon,stratagem} \
```
[the remark this answers is not in the document]
Comment (jgrund) [reply .copr/Makefile]: These are all binaries or systemd units.

File: iml-agent/src/daemon_plugins/daemon_plugin.rs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+/// Maintains internal state and sends and receives messages.
+///
+/// Implementors of this trait should add themselves
+/// to the `plugin_registry` below.
+pub trait DaemonPlugin: std::fmt::Debug {
+    /// Returns full listing of information upon session esablishment
+    fn start_session(&self) -> Box<Future<Item = Output, Error = ImlAgentError> + Send> {
+        Box::new(future::ok(None))
+    }
+    ///  Return information needed to maintain a manager-agent session, i.e. what
```
[the remark this answers is not in the document]
Comment (jgrund) [reply iml-agent/src/daemon_plugins/daemon_plugin.rs]: Was this an accident? not sure what the suggested change should be.

File: iml-agent/src/daemon_plugins/daemon_plugin.rs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    /// has changed since the start of the session or since the last update.
+    ///
+    /// If you need to refer to any data from the start_session call, you can
+    /// store it as property on this DaemonPlugin instance.
```
[the remark this answers is not in the document]
Comment (jgrund) [reply iml-agent/src/daemon_plugins/daemon_plugin.rs]: Was this an accident? not sure what the suggested change should be.

File: python-iml-manager.spec.in (hunk 1 above)
Comment (jgrund) [reply python-iml-manager.spec.in]: yeah. The rust-iml.spec in this patch does use copr builder.

File: iml-agent/src/daemon_plugins/daemon_plugin.rs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+pub fn plugin_registry() -> DaemonPlugins {
+    let mut hm = HashMap::new();
+
+    hm.insert(
+        PluginName("stratagem".into()),
+        mk_callback(&stratagem::create),
+    );
+
+    // @FIXME: This should be action_runner. 2 at the end is
+    // a temporary workaround to get response side working consistently.
```
[the remark this answers is not in the document]
Comment (jgrund) [reply iml-agent/src/daemon_plugins/daemon_plugin.rs]: No ticket, I should probably make this action_runner and setup the queues to be differentiated on the manager side. 

I can do that in a follow-on patch if you're ok with that.

File: iml-agent/src/env.rs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    env::var(name).unwrap_or_else(|_| panic!("{} environment variable is required.", name))
+}
+
+/// Gets the manager url or panics
+lazy_static! {
+    pub static ref MANAGER_URL: Url =
+        Url::parse(&get_var("IML_MANAGER_URL")).expect("Could not parse manager url");
+}
+
+static PRIVATE_PEM_PATH: &str = "/etc/iml/private.pem";
```
[the remark this answers is not in the document]
Comment (jgrund) [reply iml-agent/src/env.rs]: This is lazyness on my part, it should be an environment variable like the rest. I'll fix.

File: iml-agent/src/daemon_plugins/daemon_plugin.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (johnsonw) [reply iml-agent/src/daemon_plugins/daemon_plugin.rs]: There is an extra space in between "///" and "Return"

File: iml-agent/src/daemon_plugins/daemon_plugin.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (johnsonw) [reply iml-agent/src/daemon_plugins/daemon_plugin.rs]: Just suggesting to put "a" in between "as" and "property"

File: iml-agent/src/daemon_plugins/daemon_plugin.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (johnsonw) [reply iml-agent/src/daemon_plugins/daemon_plugin.rs]: fine with me

File: iml-agent/src/daemon_plugins/daemon_plugin.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jgrund) [reply iml-agent/src/daemon_plugins/daemon_plugin.rs]: Got it.

File: iml-agent/src/env.rs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+                    "pkcs12",
+                    "-export",
+                    "-out",
+                    PFX_PATH,
+                    "-inkey",
+                    PRIVATE_PEM_PATH,
+                    "-in",
+                    CRT_PATH,
+                    "-certfile",
+                    "/etc/iml/authority.crt",
```
[the remark this answers is not in the document]
Comment (jgrund) [reply iml-agent/src/env.rs]: Yes, will fix.

**Pull request merged**